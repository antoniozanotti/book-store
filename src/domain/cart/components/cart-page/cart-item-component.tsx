import { CartItemType } from "../../types/cart-item-type";
import { UiLoaderComponent } from "@/domain/ui/components/ui-loader-component";
import { useProductByIdQuery } from "@/domain/product/queries/use-product-by-id-query";
import { ProductImage } from "./product-image-component";
import { ProductDescription } from "./product-description-component";
import { CartItemActions } from "./cart-item-actions-component";

export function CartItemComponent({ cartItem }: { cartItem: CartItemType }) {
  const { isLoading, data } = useProductByIdQuery(cartItem.id);

  if (isLoading) return <UiLoaderComponent />;
  if (!data) return null;

  const product = data;

  return (
    <li className="py-4 border-b grid grid-cols-[145px_calc(100%-145px-16px)] gap-4">
      <ProductImage product={product} />
      <ProductDescription product={product} />
      <CartItemActions product={product} cartItemQty={cartItem.quantity} />
    </li>
  );
}
