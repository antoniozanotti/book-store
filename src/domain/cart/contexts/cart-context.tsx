"use client";

import {
  createContext,
  useContext,
  useReducer,
  ReactNode,
  useEffect,
} from "react";
import { CartType } from "../types/cart-type";
import { CartContextType } from "../types/cart-context-type";
import { useLocalStorage } from "usehooks-ts";
import { useProductsByIdsQuery } from "@/domain/product/queries/use-products-by-ids-query";
import { cartReducer } from "../reducers/cart-reducer";

const initialState: CartType = { items: [] };

const CartContext = createContext<CartContextType>({
  state: initialState,
  dispatch: () => {},
  getCartItemsQty: () => {},
  cartItemsSubtotal: () => {},
});

export function CartProvider({ children }: { children: ReactNode }) {
  const [localState, setLocalState] = useLocalStorage<CartType>(
    "cart",
    initialState
  );
  const [state, dispatch] = useReducer(cartReducer, localState);
  const cartItems = state.items;
  const responses = useProductsByIdsQuery(cartItems.map((item) => item.id));

  function getCartItemsQty(): number {
    const cartItems = state.items;
    return cartItems.reduce((acc, cur) => {
      return acc + cur.quantity;
    }, 0);
  }

  function cartItemsSubtotal(): number {
    let subTotal = 0;
    cartItems.forEach((item) => {
      responses.some((response) => {
        const product = response.data;
        if (!product) {
          return true;
        }
        if (item.id == product.id) {
          subTotal += parseInt(product.price.toString()) * item.quantity;
          return true;
        }
      });
    });
    return subTotal;
  }

  useEffect(() => {
    setLocalState(state);
  }, [state, setLocalState]);

  return (
    <CartContext.Provider
      value={{ state, dispatch, getCartItemsQty, cartItemsSubtotal }}
    >
      {children}
    </CartContext.Provider>
  );
}

export function useCartContext(): CartContextType {
  return useContext(CartContext);
}
