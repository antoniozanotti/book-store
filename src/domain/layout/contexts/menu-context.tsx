"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { MenuType } from "../types/menu-type";
import { MenuContextType } from "../types/menu-context-type";
import { menuReducer } from "../reducers/menu-reducer";

const initialState: MenuType = { isOpen: false };

const MenuContext = createContext<MenuContextType>({
  state: initialState,
  dispatch: () => {},
});

export function MenuProvider({ children }: { children: ReactNode }) {
  const [state, dispatch] = useReducer(menuReducer, initialState);

  return (
    <MenuContext.Provider value={{ state, dispatch }}>
      {children}
    </MenuContext.Provider>
  );
}

export function useMenuContext(): MenuContextType {
  return useContext(MenuContext);
}
