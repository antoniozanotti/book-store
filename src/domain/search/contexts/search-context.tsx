"use client";

import { createContext, useContext, useReducer, ReactNode } from "react";
import { SearchContextType } from "../types/search-context-type";
import { useSearchParams } from "next/navigation";
import { getSearchParams } from "../services/search-service";
import { searchReducer } from "../reducers/search-reducer";

const SearchContext = createContext<SearchContextType>({} as SearchContextType);

export function SearchProvider({ children }: { children: ReactNode }) {
  const searchParams = useSearchParams();
  const { keyword, categories, minPrice, maxPrice, minYear, maxYear, orderBy } =
    getSearchParams(searchParams);
  const initialStateBySearchParams = {
    keyword: keyword,
    categories: categories,
    minprice: minPrice,
    maxprice: maxPrice,
    minyear: minYear,
    maxyear: maxYear,
    orderby: orderBy,
  };
  const [state, dispatch] = useReducer(
    searchReducer,
    initialStateBySearchParams
  );

  return (
    <SearchContext.Provider value={{ state, dispatch }}>
      {children}
    </SearchContext.Provider>
  );
}

export function useSearchContext(): SearchContextType {
  return useContext(SearchContext);
}
