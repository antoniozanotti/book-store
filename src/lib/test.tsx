import React, { ReactElement, ReactNode } from "react";
import {
  render,
  RenderOptions,
  RenderResult,
  Screen,
} from "@testing-library/react";

function AllTheProviders({ children }: { children: React.ReactNode }) {
  return <div data-testid="container">{children}</div>;
}

function customRender(
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

function firstComponent(screen: Screen) {
  return screen.getByTestId("container").firstChild;
}

export * from "@testing-library/react";
export * from "@testing-library/user-event";
export { customRender as render, firstComponent };
import "@testing-library/jest-dom";
import { usePathname } from "next/navigation";

jest.mock("next/navigation", () => {
  return {
    useRouter() {
      return { push: () => {} };
    },
    useParams: () => ({}),
    notFound: () => null,
    usePathname: () => null,
    useSearchParams: () => null,
  };
});
