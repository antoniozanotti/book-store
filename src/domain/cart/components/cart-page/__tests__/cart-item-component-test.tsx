import { render } from "@/lib/test";
import { CartItemComponent } from "../cart-item-component";
import { CartProvider } from "@/domain/cart/contexts/cart-context";
import { productWithPagEandBookMock } from "@/domain/product/__tests__/product-with-page-and-book-mock";
import { TanstackProvider } from "@/lib/tanstack-provider";

jest.mock("@/domain/product/queries/use-product-by-id-query", () => {
  const originalModule = jest.requireActual(
    "@/domain/product/queries/use-product-by-id-query"
  );
  return {
    __esModule: true,
    useProductByIdQuery: jest.fn((id: number) => {
      return {
        isLoading: false,
        data: productWithPagEandBookMock,
      };
    }),
  };
});

describe("CartItemComponent", () => {
  it("should render", async () => {
    const { container } = render(
      <TanstackProvider>
        <CartProvider>
          <CartItemComponent
            cartItem={{ id: productWithPagEandBookMock.id, quantity: 1 }}
          />
        </CartProvider>
      </TanstackProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
