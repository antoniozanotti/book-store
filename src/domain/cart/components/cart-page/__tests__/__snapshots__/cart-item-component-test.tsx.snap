// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItemComponent should render 1`] = `
<div>
  <div
    data-testid="container"
  >
    <li
      class="py-4 border-b grid grid-cols-[145px_calc(100%-145px-16px)] gap-4"
    >
      <a
        class="w-[145px] lg:row-span-2"
        href="/product/the-four-agreements-a-practical-guide-to-personal-freedom"
        title="The Four Agreements: A Practical Guide to Personal Freedom by Don Miguel Ruiz"
      >
        <figure
          class="bg-green-300 w-[145px] h-[193px] relative"
        >
          <img
            alt="The Four Agreements: A Practical Guide to Personal Freedom"
            class="transition-all object-cover w-full h-full"
            data-nimg="1"
            decoding="async"
            height="193"
            loading="lazy"
            sizes="145px"
            src="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5111,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0"
            srcset="https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_16,h_21,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 16w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_32,h_42,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 32w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_48,h_63,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 48w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_64,h_85,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 64w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_96,h_127,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 96w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_128,h_170,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 128w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_256,h_340,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 256w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_384,h_511,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 384w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_640,h_851,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 640w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_750,h_998,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 750w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_828,h_1102,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 828w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1080,h_1437,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 1080w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1200,h_1597,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 1200w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_1920,h_2555,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 1920w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_2048,h_2725,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 2048w, https://res.cloudinary.com/dulz2k2my/image/upload/c_auto,w_3840,h_5111,g_auto/f_auto/q_auto/v1/product/the-four-agreements-a-practical-guide-to-personal-freedom?_a=BAVARSDW0 3840w"
            style="color: transparent;"
            width="145"
          />
        </figure>
      </a>
      <div
        class="h-fit"
      >
        <a
          class="line-clamp-2"
          href="/product/the-four-agreements-a-practical-guide-to-personal-freedom"
          title="The Four Agreements: A Practical Guide to Personal Freedom by Don Miguel Ruiz"
        >
          <h2>
            The Four Agreements: A Practical Guide to Personal Freedom
          </h2>
        </a>
        <span
          class="line-clamp-2 block text-sm"
        >
          by 
          Don Miguel Ruiz
        </span>
        <span
          class="block text-sm my-2"
        >
          ISBN: 
          978-1-315-92574-5
        </span>
      </div>
      <div
        class="flex gap-2 h-fit self-end col-span-2 lg:col-auto"
      >
        <button
          class="bg-primary-900 font-bold p-4 text-primary-100 text-sm p-0 bg-transparent text-red-700"
        >
          remove
        </button>
        <div
          class="border-y border-primary-900 flex gap-2 items-center"
        >
          <button
            aria-label="decrease quantity"
            class="bg-primary-900 font-bold p-4 text-primary-100 text-sm p-1.5"
          >
            <svg
              class="lucide lucide-minus "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 12h14"
              />
            </svg>
          </button>
          <div
            aria-label="quantity"
            class="w-10 text-center text-sm"
          >
            1
          </div>
          <button
            aria-label="increase quantity"
            class="bg-primary-900 font-bold p-4 text-primary-100 text-sm p-1.5"
          >
            <svg
              class="lucide lucide-plus "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 12h14"
              />
              <path
                d="M12 5v14"
              />
            </svg>
          </button>
        </div>
        <span
          class="font-bold grow flex items-end justify-end"
        >
          <span
            aria-label="cart item subtotal"
          >
            $302.00
          </span>
        </span>
      </div>
    </li>
  </div>
</div>
`;
